src/simple_graph.py

```diff
@@ -0,0 +1,82 @@
+"""An implementation of a simple graph."""
+
+
+class Graph(object):
+    """A graph class.
+
+    nodes() - Return list of nodes in graph
+
+    edges() - Return all edges in graph
+
+    add_node(n) -add node n to graph
+
+    add_edge(n1. n2) - add edge between n1 and n2 add nodes if not already in graph
+
+    del_node(n) - delete node n
+
+    del_edge(n1, n2) - del edge between n1 and n2
+
+    has_node(n) - check if node n is in graph
+
+    neighbors(n) - return list of neighbors of n
+
+    adjacent(n1, n2) - check if n1 and n2 are adjacent.
+
+    """
+
+    def __init__(self):
+        """Initialize a graph with a dict."""
+        self._nodes = {}
+
+    def nodes(self):
+        """Return list of nodes."""
+        return list(self._nodes.keys())
+
+    def edges(self):
+        """Return list of edges."""
+        return list({key + '-' + c for key, val in self._nodes.items() for c in ''.join(val)})
+
+    def add_node(self, n):
+        """Add a node to the graph."""
+        if not self.has_node(n):
+            self._nodes[n] = []
+
+    def add_edge(self, n1, n2):
+        """Add edge to graph, if nodes not in graph, add them."""
+        if self.has_node(n1):
+            self._nodes[n1].append(n2)
+        else:
+            self._nodes.update({n1:[n2]})
```

Comment:
instead of an if/else, you could call your add_node method which will only add the node if it doesn't exist already then use self._nodes[n1].append(n2).